sw/device/tests/aon_timer_wakeup_irq_test.c

```diff
@@ -0,0 +1,191 @@
+
+  CHECK_DIF_OK(dif_rv_timer_reset(&rv_timer));
+
+  // Compute and set tick parameters (i.e., step, prescale, etc.).
+  dif_rv_timer_tick_params_t tick_params;
+  CHECK_DIF_OK(dif_rv_timer_approximate_tick_params(kClockFreqPeripheralHz,
+                                                    kTickFreqHz, &tick_params));
+
+  CHECK_DIF_OK(
+      dif_rv_timer_set_tick_params(&rv_timer, kPlicTarget, tick_params));
+
+  CHECK_DIF_OK(dif_rv_timer_counter_set_enabled(&rv_timer, kPlicTarget,
+                                                kDifToggleEnabled));
+}
+
+/**
+ * Read the current rv timer count/tick.
+ *
+ * @return The current rv timer count.
+ */
+static uint64_t tick_count_get(void) {
+  uint64_t tick = 0;
+  CHECK_DIF_OK(dif_rv_timer_counter_read(&rv_timer, kPlicTarget, &tick));
+  return tick;
+}
+
+/**
+ * Execute the wake up interrupt test.
+ */
+static void execute_test(dif_aon_timer_t *aon_timer, uint32_t sleep_time_us) {
+  peripheral = kTopEarlgreyPlicPeripheralUnknown;
+  irq = kDifAonTimerIrqWdogTimerBark;
+  // The wake up time should be `sleepTime_us ±5%`
+  uint32_t variation = sleep_time_us * 5 / 100;
+  CHECK(variation > 0);
+  uint32_t sleep_range_h = sleep_time_us + variation;
+  uint32_t sleep_range_l = sleep_time_us - variation;
+
+  // Setup the timer and wait for the IRQ
+  uint32_t sleep_cycles = (sleep_time_us * kClockFreqAonHz / 1000000);
+  LOG_INFO("Setting wakeup interrupt for %u us (%u cycles)", sleep_time_us,
+           sleep_cycles);
+  // Capture the current tick to measure the time the IRQ will take
+  time_elapsed = tick_count_get();
+  aon_timer_testutils_wakeup_config(aon_timer, sleep_cycles);
+  do {
+    wait_for_interrupt();
+  } while (peripheral != kTopEarlgreyPlicPeripheralAonTimerAon &&
+           time_elapsed < sleep_range_h);
+
+  CHECK(time_elapsed < sleep_range_h && time_elapsed > sleep_range_l,
+        "Timer took %u usec which is not in the range %u usec and %u usec",
+        (uint32_t)time_elapsed, sleep_range_l, sleep_range_h);
+
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralAonTimerAon,
+        "Interrupt from incorrect peripheral: exp = %d, obs = %d",
+        kTopEarlgreyPlicPeripheralAonTimerAon, peripheral);
+
+  CHECK(irq == kDifAonTimerIrqWkupTimerExpired,
+        "Interrupt type incorrect: exp = %d, obs = %d",
+        kDifAonTimerIrqWkupTimerExpired, irq);
+}
+
+/**
+ * External interrupt handler.
+ */
+void ottf_external_isr(void) {
+  // Calc the elapsed time since the activation of the IRQ
+  time_elapsed = tick_count_get() - time_elapsed;
+
+  dif_rv_plic_irq_id_t irq_id;
+  CHECK_DIF_OK(dif_rv_plic_irq_claim(&plic, kPlicTarget, &irq_id));
+
+  peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[irq_id];
+
+  if (peripheral == kTopEarlgreyPlicPeripheralAonTimerAon) {
+    irq = (dif_aon_timer_irq_t)(
+        irq_id -
+        (dif_rv_plic_irq_id_t)kTopEarlgreyPlicIrqIdAonTimerAonWkupTimerExpired);
```

Comment:
This kind of mapping could potentially be useful in a testutil. Perhaps it could be a generic helper to convert any plic IRQ ID to the peripheral IRQ ID (assumption that the peripheral local IRQ order is the same as the PLIC IRQ ID order for this peripheral must be valid).

This is, however, outside of the scope of this PR.

Reply:
I think there should be a testutil to parse all the possible interrupts and acknowledge them and then call callback functions from the test to handle specific interrupts.

Reply:
@silvestrst, I think what you suggest is tracked in #9038.